configs/provider_requirements.go

```diff
@@ -26,132 +29,214 @@ type RequiredProviders struct {
-					} else {
-						rp.Type = fqn
 					}
+					diags = append(diags, hclDiags...)
+					continue
 				}
-			}
-			attrTypes := expr.Type().AttributeTypes()
-			for name := range attrTypes {
-				if name == "version" || name == "source" {
+
+				rp.Source = source.AsString()
+				rp.Type = fqn
+
+			case "configuration_aliases":
+				exprs, listDiags := hcl.ExprList(kv.Value)
+				if listDiags.HasErrors() {
+					diags = append(diags, listDiags...)
 					continue
 				}
+
+				for _, expr := range exprs {
+					traversal, travDiags := hcl.AbsTraversalForExpr(expr)
+					if travDiags.HasErrors() {
+						diags = append(diags, travDiags...)
+						continue
+					}
+
+					addr, cfgDiags := ParseProviderConfigCompact(traversal)
+					if cfgDiags.HasErrors() {
+						diags = append(diags, &hcl.Diagnostic{
+							Severity: hcl.DiagError,
+							Summary:  "Invalid configuration_aliases value",
+							Detail:   `Configuration aliases can only contain references to local provider configuration names in the format of provider.alias`,
+							Subject:  kv.Value.Range().Ptr(),
+						})
+						continue
+					}
+
+					if addr.LocalName != name {
+						diags = append(diags, &hcl.Diagnostic{
+							Severity: hcl.DiagError,
+							Summary:  "Invalid configuration_aliases value",
+							Detail:   fmt.Sprintf(`Configuration aliases must be prefixed with the provider name. Expected %q, but found %q.`, name, addr.LocalName),
+							Subject:  kv.Value.Range().Ptr(),
+						})
+						continue
+					}
+
+					rp.Aliases = append(rp.Aliases, addr)
+				}
+
+			default:
 				diags = append(diags, &hcl.Diagnostic{
 					Severity: hcl.DiagError,
 					Summary:  "Invalid required_providers object",
-					Detail:   `required_providers objects can only contain "version" and "source" attributes. To configure a provider, use a "provider" block.`,
-					Subject:  attr.Expr.Range().Ptr(),
+					Detail:   `required_providers objects can only contain "version", "source" and "configuration_aliases" attributes. To configure a provider, use a "provider" block.`,
+					Subject:  kv.Key.Range().Ptr(),
 				})
 				break
 			}
 
-		default:
-			// should not happen
-			diags = append(diags, &hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid required_providers syntax",
-				Detail:   "required_providers entries must be strings or objects.",
-				Subject:  attr.Expr.Range().Ptr(),
-			})
 		}
 
-		if rp.Type.IsZero() && !diags.HasErrors() { // Don't try to generate an FQN if we've encountered errors
-			pType, err := addrs.ParseProviderPart(rp.Name)
-			if err != nil {
-				diags = append(diags, &hcl.Diagnostic{
-					Severity: hcl.DiagError,
-					Summary:  "Invalid provider name",
-					Detail:   err.Error(),
-					Subject:  attr.Expr.Range().Ptr(),
-				})
-			} else {
-				rp.Type = addrs.ImpliedProviderForUnqualifiedType(pType)
+		// finally add the required provider as long as there were no errors
+		if !diags.HasErrors() {
```

Comment:
I have no objection, but I'm curious why you split this if statement into two?

Reply:
Oh! It used to have more there, but you're right, the logic can now be collapsed.

Reply:
Looking at the expanded file I see why, there is an assignment outside of the inner `if` block. Since I'm updating the file anyway, I'll move the first condition out completely.